chore: Dependency 변경 - JDBC삭제, JPA등록 / feat: admin 메인화면, 카테고리 추가, 수정 삭제 API 구현

HAKSUYOON opened the pull request:
### 변경사항
<!-- 이 PR에서 어떤점들이 변경되었는지 기술해주세요. 가급적이면 as-is, to-be를 활용해서 작성해주세요.  -->
**AS-IS**
- Dependency 변경 필요 : JDBC 삭제, JPA 추가
- Beverage 등록을 위해 Category 기능이 우선적으로 필요
**TO-BE**
- Dependency 변경 완료 : JDBC 삭제, JPA 추가
- Category API 구현 : 추가, 삭제, 수정 API 구현 (임시 ADMIN 페이지 작성)
### 테스트
<!-- 본 변경사항이 테스트가 되었는지 기술해주세요 --> 
- [ ] 테스트 코드
- [ ] API 테스트

HAKSUYOON committed:
chore: Dependency 변경 - JDBC삭제, JPA등록
feat: admin 메인화면, 카테고리 추가, 수정 삭제 API 구현

Seokjin-K reviewed (commented):
한눈에 봐도 어떤 의도를 가진 코드인지 알 수 있는 깔끔한 코드네요. 저도 이런 깔끔하고 간결한 코딩을 배워야겠습니다. 프로젝트 기간 동안 잘 부탁드립니다.

mingming-mentor commented on src/main/java/com/zerobase/cafekiosk/admin/category/controller/CategoryController.java:
```diff
@@ -0,0 +1,52 @@
+
+  private final CategoryService categoryService;
+
+  @GetMapping("/admin/category/list.do")
+  public String list(Model model) {
+
+    List<CategoryDto> categoryDtoList = categoryService.list();
+    model.addAttribute("list", categoryDtoList);
+
+    return "admin/category/list";
+  }
```
전체적으로 Restful API 개발이 아닌 페이지 + model 형태로 개발을하신 것 같은데요, Restful api 개발을 하지 않으신 이유가 있을까요?

mingming-mentor replied:
Restful API 형태로 개발을 하신다면 url convention이 존재를 합니다. 이 부분도 참고해보심 좋을 것 같네요 
https://restfulapi.net/resource-naming/

mingming-mentor commented on src/main/java/com/zerobase/cafekiosk/admin/category/dto/CategoryDto.java:
```diff
@@ -0,0 +1,42 @@
+      }
+      return categoryDtoList;
+    }
+    return null;
```
categories가 empty일 경우 null이 반환이 되는 것같아요. 동일하게 empty list가 반환되는 것이 아니라 null을 반환시키는 이유가 있을까요? NPE 문제도 있기도 하고, null과 empty를 어떤 기준으로 구분해서 사용하시는지 궁금하네요

mingming-mentor commented on src/main/java/com/zerobase/cafekiosk/admin/category/dto/CategoryDto.java:
```diff
@@ -0,0 +1,42 @@
+
+  public static List<CategoryDto> of(List<Category> categories) {
+
+    if (!categories.isEmpty()) {
+      List<CategoryDto> categoryDtoList = new ArrayList<>();
+      for (Category x : categories) {
+        categoryDtoList.add(of(x));
+      }
+      return categoryDtoList;
+    }
```
stream을 쓰면 좀더 간결하게 작성 가능할 듯 합니다

mingming-mentor commented on src/main/java/com/zerobase/cafekiosk/admin/category/dto/CategoryDto.java:
```diff
@@ -0,0 +1,42 @@
+@NoArgsConstructor
+public class CategoryDto {
+
+  Long id;
+  String categoryName;
+  int sortValue;
+  boolean usingYn;
```
앞에 private을 붙이지 않으신 이유가 있나요?

mingming-mentor commented on src/main/java/com/zerobase/cafekiosk/admin/category/service/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,70 @@
+  @Override
+  public boolean add(String categoryName) {
+
+    Optional<Category> optionalCategory = categoryRepository.findByCategoryName(categoryName);
+    if (optionalCategory.isPresent()) {
+      return false;
+    }
```
optionalCategory를 별도로 선언할 이유가 없어보입니다
```suggestion
    if (categoryRepository.findByCategoryName(categoryName).isPresent()) {
      return false;
    }
```

mingming-mentor replied:
이미 존재하는 카테고리명으로 저장을 시도할때, 저장이 안되는데요. 사용자에게 이에 대해서 어떤 이유때문에 안되는지를 알려줄 필요는 없나요? 단순히 false로 return을 해버리면 추후 다른 비즈니스로직이 들어와서 add 가 되지 않을때도 동일하게 false, 즉 유저는 어떤 이유때문에 add가 안된것인지 알기가 어렵습니다.
보통 이런경우 exception을 던지고 GlobalExceptionHandler에서 이를 처리하도록 하면서 exception의 reason도 같이 넣어주는 형태로 개발되곤 합니다

mingming-mentor commented on src/main/java/com/zerobase/cafekiosk/admin/category/service/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,70 @@
+      category.setUsingYn(parameter.isUsingYn());
+      categoryRepository.save(category);
+    }
+    return true;
```
성공시에는 굳이 true를 던져줄 필요가 없을 것 같아요. 위의 코멘트와 동일한 맥락인데
성공시에는 void, 실패시에는 exception을 던져주는 형태가 더 깔끔해보입니다. 
그리고 만약 업데이트하고자 한 id가 존재하지 않는 경우는 별도로 exception을 던져주지 않고 true로 처리하는데 의도하신바가 맞나요?

mingming-mentor commented on src/main/java/com/zerobase/cafekiosk/admin/category/service/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,70 @@
+  @Override
+  public boolean update(CategoryInput parameter) {
+
+    Optional<Category> optionalCategory = categoryRepository.findById(parameter.getId());
+    if (optionalCategory.isPresent()) {
```
여기서도 optionalCategory는 선언하지 않는 것이 가독성 측면에서, 또 optional을 사용하는 측면에서 더 낫겠습니다
(라이브 QnA때 드린 설명과 같은 맥락입니다)

mingming-mentor reviewed (approved):
리뷰드렸습니다

HAKSUYOON committed:
chore: Dependency 변경 - Thymeleaf 삭제
feat: Restful API로 구현방식 변경, 사용자에게 보여지는 frontList Service 추가 (ADMIN 권한의 기능이 아니므로 나중에 구현)
